### ISSUE TITLE
Shaping Thai text support

### ISSUE THREAD

**Issue by mrchoke** [py-pdf/fpdf2#679]
**Please explain your intent**
Now fpdf2 can display Thai text but it not pretty. 

**Describe the solution you'd like**
Reference for shaping thai text.

- https://github.com/n8willis/opentype-shaping-documents/blob/master/opentype-shaping-thai-lao.md
- https://linux.thai.net/~thep/th-otf/shaping.html
- https://linux.thai.net/~thep/th-otf/
- https://www.nectec.or.th/it-standards/thaistd.pdf

**Additional context**
![Screenshot 2023-02-04 at 15 19 22](https://user-images.githubusercontent.com/3403645/216757024-b95fccee-5bab-4e37-8703-e94795cb541c.jpg)

**Comment by andersonhc** [comment]
@mrchoke can you please give the Thai text so I can copy/paste it? I am doing some tests integrating harfbuzz on fpdf2 and I'd like to know if it solves your problem too.

**Comment by mrchoke** [comment]
> https://pyfpdf.github.io/fpdf2/Unicode.html



> Have you read this section from the documentation?
> 
> >

Yes, I read it before. Now FPDF2 can display thai text at sample level. Thai writing system too complex.
I used WeasyPrint pdf generator it use Pango for thai text rendering it so pretty and it can manage thai word/line break too.

**Comment by mrchoke** [comment]
> @mrchoke can you please give the Thai text so I can copy/paste it? I am doing some tests integrating harfbuzz on fpdf2 and I'd like to know if it solves your problem too.

Good news @andersonhc.

ok, wait me a while, I'll do it for you as soon as possible.

**Comment by mrchoke** [comment]
## Sample Thai text:

- ป ปี ปี่ ปี้ ปี๊ ปี๋ ปา ป่า ป้า ป๊า ป๋า ปำ ป่ำ ปุ ปู ปฺ ปํ ปิํ
- บ บี บี่ บี้่ บี๊ บี๋ บา บ่า บ้า บ๊า บ๋า บำ บ่ำ บุ บู บฺ บํ บิํ
- ฬ ฬี ฬี่ ฬี้ ฬี๊ ฬี๋ ฬา ฬ่า ฬ้า ฬ๊า ฬ๋า ฬำ ฬ่ำ ฬุ ฬู ฬฺ ฬํ ฬิํ
-  ญ ญี ญี่ ญี้ ญี๊ ญี๋ ญา ญ่า ญ้า ญ๊า ญ๋า ญำ ญ่ำ ญุ ญู ญฺ ญํ ญิํ
-  ฐ ฐี ฐี่ ฐี้ ฐี๊ ฐี๋ ฐา ฐ่า ฐ้า ฐ๊า ฐ๋า ฐำ ฐ่ำ ฐุ ฐู ฐฺ ฐํ ฐิํ
- ฎ ฎี ฎี่ ฎี้ ฎี๊ ฎี๋ ฎา ฎ่า ฎ้า ฎ๊า ฎ๋า ฎำ ฎ่ำ ฎุ ฎู ฎฺ ฎํ ฎิํ
- ฏ ฏี ฏี่ ฏี้ ฏี๊ ฏี๋ ฏา ฏ่า ฏ้า ฏ๊า ฏ๋า ฏำ ฏ่ำ ฏุ ฏู ฏฺ ฏํ ฏิํ

## Current FPDF2 Render:
![Screenshot 2023-02-08 at 01 28 05](https://user-images.githubusercontent.com/3403645/217334085-f6e38e77-70e9-41cb-972c-e0774e164351.jpg)

## Expected:
![Screenshot 2023-02-08 at 01 29 50](https://user-images.githubusercontent.com/3403645/217334111-03701653-ac26-4fa5-94f3-7d6e24475f3c.jpg)

Reference:
https://www.nectec.or.th/it-standards/thaistd.pdf
Page 5
![Screenshot 2023-02-07 at 23 12 11](https://user-images.githubusercontent.com/3403645/217333294-9bd107c1-7a3f-4089-81ad-8c2125431f96.jpg)
![Screenshot 2023-02-07 at 23 12 46](https://user-images.githubusercontent.com/3403645/217333317-092381a5-aea7-4bb5-bd2f-369ff4c1c1df.jpg)


@andersonhc

**Comment by Lucas-C** [comment]
> @mrchoke can you please give the Thai text so I can copy/paste it? I am doing some tests integrating harfbuzz on fpdf2 and I'd like to know if it solves your problem too.

Wow, that sounds promising!
I'd be glad to know how those tests went: what kind of problems have you faced?
Are there missing bricks in `fpdf2` to plug it with `harfbuzz`?

**Comment by eroux** [comment]
I do have an example with Tibetan, using the [NotoSerifTibetan](https://fonts.google.com/noto/specimen/Noto+Serif+Tibetan?noto.lang=bo_Tibt&noto.continent=Asia&noto.script=Tibt) font:

```python
from fpdf import FPDF

pdf = FPDF()
pdf.add_page(format=(200,200))
pdf.add_font("NotoSerifTibetan", fname='NotoSerifTibetan-Regular.ttf')
pdf.set_font("NotoSerifTibetan", '', 56)
pdf.write(40, "༄༅། །སྒྲུབ།")
pdf.output("test-tib.pdf")
```

result :

[test-tib.pdf](https://github.com/PyFPDF/fpdf2/files/10732707/test-tib.pdf)

expected result (produced with LibreOffice Writer):

[test-tib-lo.pdf](https://github.com/PyFPDF/fpdf2/files/10732710/test-tib-lo.pdf)

**Comment by andersonhc** [comment]
> > @mrchoke can you please give the Thai text so I can copy/paste it? I am doing some tests integrating harfbuzz on fpdf2 and I'd like to know if it solves your problem too.
> 
> Wow, that sounds promising! I'd be glad to know how those tests went: what kind of problems have you faced? Are there missing bricks in `fpdf2` to plug it with `harfbuzz`?

My biggest problem is that fpdf2 maps 1 unicode character to 1 glyph on the font, and that doesn't work with font shapers that supports ligatures (multiple characters will render 1 glyph).

I changed the SubsetMap to work with Glyphs and did several changes on the "font" code (using classes instead of dictionary).

I have some more things to fix, but as soon as I get a working prototype version I'll open a discussion.

**Comment by Lucas-C** [comment]
> I changed the SubsetMap to work with Glyphs and did several changes on the "font" code (using classes instead of dictionary).
> I have some more things to fix, but as soon as I get a working prototype version I'll open a discussion.

Awesome! Best of luck with this project 😊
Your contribution would be very welcome.

**Comment by Lucas-C** [comment]
@andersonhc PR https://github.com/PyFPDF/fpdf2/pull/820 has been merged today.

Could you test if that solved your issue @mrchoke?

You can install `fpdf2` directly from the `master` branch of this repo with this command:

    pip install git+https://github.com/PyFPDF/fpdf2.git@master

**Comment by Lucas-C** [comment]
Since this has been released in [2.7.5](https://github.com/py-pdf/fpdf2/releases/tag/2.7.5), I'm going to close this issue.

Anyone is free to reopen it or open another one thare are still issues with thai text
